controllers/keystoneapi_controller.go

```diff
@@ -646,6 +657,18 @@ func (r *KeystoneAPIReconciler) reconcileNormal(ctx context.Context, instance *k
 		}
 	}
 
+	err = r.removeFinalizerFromOldKeystoneMemcached(ctx, helper, instance)
+	if err != nil {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			keystonev1.KeystoneMemcachedReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			keystonev1.KeystoneMemcachedReadyErrorMessage,
+			err.Error()))
+		return ctrl.Result{}, err
+	}
+	instance.Status.MemcachedInstance = instance.Spec.MemcachedInstance
+
```

Comment:
In the following unlikely sequence of events I think this code will leak memcached finalizers:
1. Spec.MemcachedInstance=old Status.MemcachedInstance=old and everything is Ready and stable
2. user set Spec.MemcachedInstance=new
3. Reconcile#1 start with Spec.MemcachedInstance=new Status.MemcachedInstance=old
4. Reconcile#1 adds finalizer to MemcachedInstance=new L646
5. Reconcile#1 removeFinalizerFromOldKeystoneMemcached() reads MemcachedInstance=old from etcd 
6. user removes the finalizer and deletes MemcachedInstance=old manually
7. Reconcile#1 removeFinalizerFromOldKeystoneMemcached() tries to update MemcachedInstance=old to remove the finalizer and gets NotFound
8. NotFound propagated to L668 and Reconcile#1 exists ( Spec.MemcachedInstance=old Status.MemcachedInstance=new is the persisted state) 
9. user updates Spec.MemcachedInstance=new2
10. Reconcile#2 starts and reads instance from etcd with Spec.MemcachedInstance=new2 Status.MemcachedInstance=old

At this point MemcachedInstance=new has our finalizer but it is not recorded in Status.MemcachedInstance so that finalizer is never removed by this operator.
This needs two precisely timed user interaction at 6) and 9) so this is an unlikely sequence.

Reply:
There is another variant of the above. If at 6) the something updates MemcachedInstance=old and therefore bumps the resource generation of it. Then at 7) we get Conflict and if the rest of the sequence is kept as is then we still end up leaking the finalizer on MemcachedInstance=new. While the NotFound case can be easily fixed by not treating NotFound as an error at 7). The Conflict scenario cannot be similarly fixed.

Reply:
Ohh. Good question. If we remove the finalizer from the old memcached first, then we have to make sure that at that point the keystone openstack service is already reconfigured to use the new memcached instance as the old can disappear. I haven't traced this case through yet...

Reply:
I eventually decided to go through all Memcached instances... that would be more robust and would save us from considering all corner cases.

Also, it allows us to do cleanup after all resources are reconciled, to ensure the memcached instance used by the current config file is kept before config files are updated and services are restarted.